mock-pop-mining/src/main/java/org/veriblock/sdk/mock/VeriBlockPopMiner.java

```diff
@@ -0,0 +1,169 @@
+        buffer.flip();
+
+        byte[] payoutInfo = new byte[80];
+        buffer.get(payoutInfo);
+
+        return payoutInfo;
+    }
+
+    BitcoinTransaction createBitcoinTx(VeriBlockBlock publishedBlock, Address address) {
+        byte[] publicationData = createPublicationData(publishedBlock, address);
+        return new BitcoinTransaction(publicationData);
+    }
+
+    // retrieve the blocks between lastKnownBlock and getChainHead()
+    List<BitcoinBlock> createBitcoinContext(BitcoinBlock lastKnownBlock) throws SQLException {
+        List<BitcoinBlock> context = new ArrayList<>();
+
+        BitcoinBlock prevBlock = bitcoinBlockchain.get(bitcoinBlockchain.getChainHead().getPreviousBlock());
+        while (prevBlock != null && !prevBlock.equals(lastKnownBlock)) {
+            context.add(prevBlock);
+            prevBlock = bitcoinBlockchain.get(prevBlock.getPreviousBlock());
+        }
+
+        Collections.reverse(context);
+        return context;
+    }
+
+    // retrieve the blocks between lastKnownBlock and getChainHead()
+    List<VeriBlockBlock> createVeriBlockContext(VeriBlockBlock lastKnownBlock) throws SQLException {
+        List<VeriBlockBlock> context = new ArrayList<>();
+
+        VeriBlockBlock prevBlock = veriblockBlockchain.get(veriblockBlockchain.getChainHead().getPreviousBlock());
+        while (prevBlock != null && !prevBlock.equals(lastKnownBlock)) {
+            context.add(prevBlock);
+            prevBlock = veriblockBlockchain.get(prevBlock.getPreviousBlock());
+        }
+
+        Collections.reverse(context);
+        return context;
+    }
+
+    Address deriveAddress(PublicKey key) {
+        String data = "V" + Base58.encode(Sha256Hash.of(key.getEncoded()).getBytes()).substring(0, 24);
+        
+        Sha256Hash hash = Sha256Hash.of(data.getBytes(StandardCharsets.UTF_8));
+        String checksum = Base58.encode(hash.getBytes()).substring(0, 4 + 1);
+
+        return new Address(data + checksum);
+    
+    }
+    
+    VeriBlockPoPTransaction signTransaction(VeriBlockPoPTransaction tx, PrivateKey privateKey) throws SignatureException, InvalidKeyException, NoSuchAlgorithmException {        
+        byte[] signature = Utils.signMessageWithPrivateKey(SerializeDeserializeService.getId(tx).getBytes(),
+                                                           privateKey);
+
+        return new VeriBlockPoPTransaction(
+                tx.getAddress(),
+                tx.getPublishedBlock(),
+                tx.getBitcoinTransaction(),
+                tx.getMerklePath(),
+                tx.getBlockOfProof(),
+                tx.getBlockOfProofContext(),
+                signature,
+                tx.getPublicKey(),
+                tx.getNetworkByte());
+    }
+
+    public VeriBlockPublication mine(VeriBlockBlock blockToEndorse, VeriBlockBlock lastKnownVBKBlock, BitcoinBlock lastKnownBTCBlock, KeyPair key) throws SQLException, SignatureException, InvalidKeyException, NoSuchAlgorithmException {
```

Comment:
Maybe just init this class with KeyPair and store it there? to not pass keypair on every call?

Reply:
I'm not yet clear as to what the final API would look like. For an attack simulation use case, ideally we want all inputs to mine() to be configurable eg several agents mining who knows what. In practice, I suspect I'll have to split this function into smaller ones as the first step and then address use cases one by one.